assertions/sync.go

```diff
@@ -0,0 +1,528 @@
+				case errors.Is(err, l2stateprovider.ErrChainCatchingUp):
+					// Otherwise, we return the error that we are still catching up to the
+					// execution state claimed by the assertion, and this function will be retried
+					// by the caller if wrapped in a retryable call.
+					chainCatchingUpCounter.Inc(1)
+					srvlog.Info("Chain still catching up to processed execution state "+
+						"will reattempt processing when caught up", log.Ctx{"err": err})
+					return false, l2stateprovider.ErrChainCatchingUp
+				case err != nil:
+					return false, err
+				}
+				return true, nil
+			})
+			if err != nil {
+				return errors.New("could not check for assertion agreements")
+			}
+			if agreedWithAssertion {
+				cursor = protocol.AssertionHash{Hash: assertion.AssertionHash}
+				m.assertionChainData.latestAgreedAssertion = cursor
+				m.assertionChainData.canonicalAssertions[cursor] = assertion
+				m.observedCanonicalAssertions <- cursor
+			}
+		}
+	}
+	return nil
+}
+
+type rivalPosterArgs struct {
+	canonicalParent  *protocol.AssertionCreatedInfo
+	invalidAssertion *protocol.AssertionCreatedInfo
+}
+
+type rivalPoster interface {
+	maybePostRivalAssertionAndChallenge(
+		ctx context.Context,
+		args rivalPosterArgs,
+	) (*protocol.AssertionCreatedInfo, error)
+}
+
+// Finds all canonical assertions from a list. Starts by setting a cursor to the
+// latest confirmed assertion, then finds all assertions parent == cursor.
+// We then check which one we agree with.
+// From there, checks all assertions that have that assertion as parent, etc.
+// This function must hold the lock on m.assertionChainData.
+func (m *Manager) respondToAnyInvalidAssertions(
+	ctx context.Context,
+	assertions []*protocol.AssertionCreatedInfo,
+	rivalPoster rivalPoster,
+) error {
+	for _, assertion := range assertions {
+		canonicalParent, hasCanonicalParent := m.assertionChainData.canonicalAssertions[protocol.AssertionHash{
+			Hash: assertion.ParentAssertionHash,
+		}]
+		_, isCanonical := m.assertionChainData.canonicalAssertions[protocol.AssertionHash{
+			Hash: assertion.AssertionHash,
+		}]
+		// If an assertion has a canonical parent but is not canonical itself,
+		// then we should challenge the assertion if we are configured to do so,
+		// or raise an alarm if we are only a watchtower validator.
+		if hasCanonicalParent && !isCanonical {
+			postedRival, err := retry.UntilSucceeds(ctx, func() (*protocol.AssertionCreatedInfo, error) {
+				posted, innerErr := rivalPoster.maybePostRivalAssertionAndChallenge(ctx, rivalPosterArgs{
+					canonicalParent:  canonicalParent,
+					invalidAssertion: assertion,
+				})
+				if innerErr != nil {
+					srvlog.Error("Could not post rival assertion and/or challenge", log.Ctx{"err": innerErr})
+					return nil, innerErr
+				}
+				return posted, nil
+			})
+			if err != nil {
+				return err
+			}
+			if postedRival != nil {
+				postedAssertionHash := protocol.AssertionHash{Hash: postedRival.AssertionHash}
```

Comment:
Something I thought about was whether or not to update the `latestAgreedAssertion` here, but it is not trivial to do so. Instead, this rival creation event will just be another assertion creation event we'll process in the following iteration of the sync code. That's the safest approach